python/hsfs/engine/spark.py

```diff
@@ -258,8 +268,49 @@ def read(self, storage_connector, data_format, read_options, path):
 
     def profile(self, dataframe, relevant_columns, correlations, histograms):
         """Profile a dataframe with Deequ."""
-        return self._jvm.com.logicalclocks.hsfs.engine.SparkEngine.getInstance().profile(
-            dataframe._jdf, relevant_columns, correlations, histograms
+        return (
+            self._jvm.com.logicalclocks.hsfs.engine.SparkEngine.getInstance().profile(
+                dataframe._jdf, relevant_columns, correlations, histograms
+            )
+        )
+
+    def validate(self, dataframe, expectations):
+        """Run data validation on the dataframe with Deequ."""
+
+        expectations_java = []
+        for expectation in expectations:
+            rules = []
+            for rule in expectation.rules:
+                rules.append(
+                    self._jvm.com.logicalclocks.hsfs.metadata.validation.Rule.builder()
+                    .name(
+                        self._jvm.com.logicalclocks.hsfs.metadata.validation.Name.valueOf(
+                            rule.get("name")
+                        )
+                    )
+                    .level(
+                        self._jvm.com.logicalclocks.hsfs.metadata.validation.Level.valueOf(
+                            rule.get("level")
+                        )
+                    )
+                    .min(rule.get("min", None))
+                    .max(rule.get("max", None))
+                    .pattern(rule.get("pattern", None))
+                    .legalValues(rule.get("legalValues", None))
+                    .build()
+                )
+            expectation = (
+                self._jvm.com.logicalclocks.hsfs.metadata.Expectation.builder()
+                .name(expectation.name)
+                .description(expectation.description)
+                .features(expectation.features)
+                .rules(rules)
+                .build()
+            )
+            expectations_java.append(expectation)
```

Comment:
Indeed it's because of what you say since it has nested complex classes
`because the Expectations cannot be passed through the gateway since it's not a standard class`

I tried to find a way to get py4j to convert it but no luck.

Reply:
too bad, but okay